### PR TITLE
fix(headless,headless-react): make ssr terminology more consistent

### PR DESCRIPTION
**Pull request by btaillon**
https://coveord.atlassian.net/browse/KIT-2680

Here's a chart of the server-side-rendering flow:
```mermaid
sequenceDiagram
  participant Headless as @coveo/headless
  participant Static as Static definition
  participant SSR as Server-side
  participant CSR as Client-side
  alt Create definition
    Static->>+Headless: const definition = defineSearchEngine({ ... });
    Headless->>-Static: engine definition
  end
  CSR->>+SSR: request page
  alt Execute the first search
    SSR->>+Static: const stateA = definition.getStateA();
    Static->>SSR: First search response
    Static->>-SSR: State of every controller
  end
  alt render page
    SSR->>SSR: const html = generateHTMLFromState(stateA.controllers);
  end
  SSR->>CSR: send html
  SSR->>-CSR: send stateA
  CSR->>CSR: display HTML
  alt Obtain engine and controllers
    CSR->>+Static: const stateB = definition.getStateB(stateA.searchResponse);
    Static->>CSR: Search engine
    Static->>-CSR: Controllers
  end
  CSR->>CSR: subscribe to controllers
```

For StateA and StateB, we've been using inconsistent terminology.

In this PR, my goal is to align all the terms to be consistent. Here are the two main options we've been discussing:
* Option 1 (option I picked): for stateA, use ~~"initial state"~~ "static state". For stateB, use "hydrated state"
  * A file shared between both server and client calls `defineSearchEngine`, which returns an `EngineDefinition<...>`
  * The server calls ~~`fetchInitialState`~~ `fetchStaticState`
    * Returns an ~~`InitialState<...>`~~ `StaticState<...>`
  * The client calls ~~`hydrateInitialState`~~ `hydrateStaticState`
    * Takes an ~~`InitialState<...>`~~ `StaticState<...>` as a parameter.
    * Returns an `HydratedState<...>`
* Option 2: for stateA, use "ssrState". For stateB, use "csrResult"
  * A file shared between both server and client calls `defineSearchEngine`, which returns an `EngineDefinition<...>`
  * The server calls `fetchSSRState`
    * Returns a `SSRState<...>`
  * The client calls `hydrateSSRState`(?)
    * Takes a `SSRState<...>` as a parameter.
    * Returns a `CSRState<...>`

I picked option 1, since it feels more natural to me after all.

### CODE REVIEW

**Review by dmbrooke — APPROVED** [review]
I like this language much more, it's very natural 🙂

**Review by olamothe — APPROVED** [review]

**Review by louis-bompart — APPROVED** [review]
LGTM, better 💯

File: packages/headless-react/src/types.ts [hunk 1]
```diff
@@ -53,9 +53,9 @@ export type ReactEngineDefinition<
...
   controllers: InferControllerHooksMapFromDefinition<TControllers>;
   useEngine(): TEngine | undefined;
-  SSRStateProvider: FunctionComponent<
+  StaticStateProvider: FunctionComponent<
     PropsWithChildren<{
-      controllers: InferControllerSSRStateMapFromDefinitions<TControllers>;
+      controllers: InferControllerStaticStateMapFromDefinitions<TControllers>;
     }>
   >;
   CSRProvider: FunctionComponent<
```
Comment (mrrajamanickam-coveo) [inline packages/headless-react/src/types.ts:61]: CSRProvider -> HydratedStateProvider ?

File: packages/headless-react/src/search-engine.test.tsx [hunk 2]
```diff
@@ -121,19 +121,19 @@ describe('Headless react SSR utils', () => {
     });
 
     test('should render with SSRProvider', async () => {
```
Comment (mrrajamanickam-coveo) [inline packages/headless-react/src/search-engine.test.tsx:123]: test desc needs to be updated

File: packages/headless/src/app/ssr-engine/types/common.ts [hunk 3]
```diff
@@ -68,15 +68,15 @@ export interface ControllerDefinitionsMap<
...
   TSearchFulfilledAction extends AnyAction,
-  TControllers extends ControllerSSRStateMap
+  TControllers extends ControllerStaticStateMap
 > {
   searchFulfilledAction: TSearchFulfilledAction;
   controllers: TControllers;
 }
 
-export interface EngineAndControllers<
+export interface HydratedState<
```
Comment (mrrajamanickam-coveo) [inline packages/headless/src/app/ssr-engine/types/common.ts:79]: Nice! Reads better.

File: packages/samples/headless-ssr/src/app/react/components/search-page.tsx [hunk 4]
```diff
@@ -1,56 +1,62 @@
...
 
 export function SearchPageProvider({
-  ssrState,
+  staticState,
   children,
 }: PropsWithChildren<SearchPageProviderProps>) {
-  const [csrResult, setCSRResult] = useState<SearchCSRState | undefined>(
-    undefined
-  );
+  const [hydratedState, setCSRResult] = useState<
```
Comment (mrrajamanickam-coveo) [inline packages/samples/headless-ssr/src/app/react/components/search-page.tsx:21]: setCSRResult -> setHydratedState

**Review by mrrajamanickam-coveo — CHANGES_REQUESTED** [review]
Reads better overall! 
Just a couple of places that got missed.

**Comment by btaillon** [comment]
Woops, missed those because I was doing a search and replace for SSR/CSR State to Static/Hydrated state.

Tried again, but this time used regular expressions like `CSR([A-Z][a-z]*)` -> `Hydrated$1` and `Static(?!State|Filter|CDN|Facet|Response|Resource)([A-Z][a-z]*)` -> `StaticState$1` to catch more. Of course, I didn't blindly approve every replacement.

**Review by mrrajamanickam-coveo — APPROVED** [review]

**Pull request merged**